tests/http/test_17_ssl_use.py

```diff
@@ -175,3 +175,68 @@ def test_17_06_localhost(self, env: Env, httpd, nghttpx, repeat, proto):
         if proto != 'h3':  # we proxy h3
             assert r.json['SSL_TLS_SNI'] == domain, f'{r.json}'
 
+    # test setting cipher suites, the AES 256 ciphers are disabled in the test server
+    @pytest.mark.parametrize("ciphers, succeed", [
+        [[0x1301], True],
+        [[0x1302], False],
+        [[0x1303], True],
+        [[0x1302, 0x1303], True],
+        [[0xC02B, 0xC02F], True],
+        [[0xC02C, 0xC030], False],
+        [[0xCCA9, 0xCCA8], True],
+        [[0xC02C, 0xC030, 0xCCA9, 0xCCA8], True],
+    ])
+    def test_17_07_ssl_ciphers(self, env: Env, httpd, nghttpx, ciphers, succeed, repeat):
+        cipher_table = {
+           0x1301: 'TLS_AES_128_GCM_SHA256',
+           0x1302: 'TLS_AES_256_GCM_SHA384',
+           0x1303: 'TLS_CHACHA20_POLY1305_SHA256',
+           0xC02B: 'ECDHE-ECDSA-AES128-GCM-SHA256',
+           0xC02F: 'ECDHE-RSA-AES128-GCM-SHA256',
+           0xC02C: 'ECDHE-ECDSA-AES256-GCM-SHA384',
+           0xC030: 'ECDHE-RSA-AES256-GCM-SHA384',
+           0xCCA9: 'ECDHE-ECDSA-CHACHA20-POLY1305',
+           0xCCA8: 'ECDHE-RSA-CHACHA20-POLY1305',
+        }
+        cipher_names = list(map(cipher_table.get, ciphers))
+        proto = 'http/1.1'
+        curl = CurlClient(env=env)
+        url = f'https://{env.authority_for(env.domain1, proto)}/curltest/sslinfo'
+        extra_args = []
+        if env.curl_uses_lib('gnutls'):
+            pytest.skip('gnutls does not support setting ciphers by name')
+        if env.curl_uses_lib('rustls-ffi'):
+            pytest.skip('rustls-ffi does not support setting ciphers')
+        if ciphers[0] & 0xFF00 == 0x1300:
+            # test setting TLSv1.3 ciphers
+            if env.curl_uses_lib('bearssl'):
+                pytest.skip('bearssl does not support TLSv1.3')
+            elif env.curl_uses_lib('sectransp'):
+                pytest.skip('sectransp does not support TLSv1.3')
+            elif env.curl_uses_lib('mbedtls'):
+                pytest.skip('mbedtls does not support TLSv1.3')
+            elif env.curl_uses_lib('boringssl'):
+                pytest.skip('boringssl does not support setting TLSv1.3 ciphers')
+            elif env.curl_uses_lib('wolfssl'):
+                extra_args = ['--ciphers', ':'.join(cipher_names)]
+            else:
+                extra_args = ['--tls13-ciphers', ':'.join(cipher_names)]
+        else:
+            # test setting TLSv1.2 ciphers
+            if env.curl_uses_lib('schannel'):
+                pytest.skip('schannel does not support setting TLSv1.2 ciphers by name')
+            elif env.curl_uses_lib('wolfssl'):
```

Comment:
Come to think of it, two remarks:

While exceptions (slips?) for ssl backends that I anticipated for are defined here, it should be mentioned that not all ssl backends are tested by pytest in CI. For example AFAIK sectranp and schannel are not.

Once mbedtls gains TLSv1.3 support, and this code is not updated, problems with setting TLSv1.3 ciphers will go on undetected. This can be anticipated by changing `pytest.skip('mbedtls does not support TLSv1.3')` to `succeed = False`, but if this is the correct/smart thing to do?

Reply:
That seems a good way to assure we detect the added support, yes.